## Association Rules

#### Association Rule Learning is rule-based learning for identifying the association between different variables in a database. One of the best and most popular examples of Association Rule Learning is the Market Basket Analysis. The problem analyses the association between various items that has the highest probability of being bought together by a customer.

### Support
Support refers to the default popularity of an item and can be calculated by finding number of transactions containing a particular item divided by total number of transactions. Suppose we want to find support for item A.

This can be calculated as:

Support(A) = (Transactions containing (A))/(Total Transactions)

For instance if out of 1000 transactions, 100 transactions contain Milk then the support for item Milk can be calculated as:

Support(Milk) = (Transactions containing Milk)/(Total Transactions)

Support(Milk) = 100/1000  
              
           = 10%

### Confidence
Confidence refers to the likelihood that an item B is also bought if item A is bought. 
It can be calculated by finding the number of transactions where A and B are bought together, 
divided by total number of transactions where A is bought. Mathematically, it can be represented as:

Confidence(A→B) = (Transactions containing both (A and B))/(Transactions containing A)  

Example: We had 50 transactions where Milk and Diaper were bought together. 
While in 150 transactions, Milk are bought. Then we can find likelihood of buying Diaper when Milk is bought can be represented as confidence of Milk -> Diaper and can be mathematically written as:

Confidence(Milk→Diaper) = (Transactions containing both (Milk and Diaper))/(Transactions containing Milk)

Confidence(Milk→Diaper) = 50/150  

                    = 33.3%

### Lift
Lift(A -> B) refers to the increase in the ratio of sale of B when A is sold. Lift(A –> B) can be calculated 
by dividing Confidence(A -> B) divided by Support(B). Mathematically it can be represented as:

Lift(A→B) = (Confidence (A→B))/(Support (B))  
Coming back to our Milk and Diaper problem, the Lift(Milk -> Diaper) can be calculated as:

Lift(Milk→Diaper) = (Confidence (Milk→Diaper))/(Support (Diaper))

Lift(Milk→Diaper) = 33.3/10  

              = 3.33
    
Lift basically tells us that the likelihood of buying Milk and Diaper together is 3.33 times more than
the likelihood of just buying the Diaper. 

A Lift of 1 means there is no association between products A and B.
Lift of greater than 1 means products A and B are more likely to be bought together. Finally, Lift of less than 
1 refers to the case where two products are unlikely to be bought together.


#### Apriori algorithm finds the most frequent itemsets or elements in a transaction database and identifies association rules between the items.

**How it helps the business:**
1. A and B can be placed together so that when a customer buys one of the product he doesn't have to go far away to buy the other product.
2. People who buy one of the products can be targeted through an advertisement campaign to buy the other.
3. Collective discounts can be offered on these products if the customer buys both of them.
4. Both A and B can be packaged together.

# Example

In [1]:
import pandas as pd
df = pd.DataFrame([[0,1,1,0,0,0],
              [1,1,0,1,1,0],
              [1,0,1,1,0,1],
              [1,1,1,1,0,0],
              [0,1,1,1,0,1]], 
             columns=['Beer', 'Bread', 'Milk', 'Diaper', 'Eggs', 'Coke'], index=['T1','T2','T3','T4','T5'])
df.head()

Beer  Bread  Milk  Diaper  Eggs  Coke
T1     0      1     1       0     0     0
T2     1      1     0       1     1     0
T3     1      0     1       1     0     1
T4     1      1     1       1     0     0
T5     0      1     1       1     0     1

In [2]:
# Install apyori (Apyori is a simple implementation of Apriori algorithm in Python)
!pip install apyori

You should consider upgrading via the 'python -m pip install --upgrade pip' command.


#### Import the Libraries

In [3]:
import numpy as np
import pandas as pd
from apyori import apriori
import matplotlib.pyplot as plt
%matplotlib inline

#### Read the Dataset

In [4]:
# store_data = pd.read_csv('store_data.csv')
store_data = pd.read_csv('store_data.csv', header=None)

#### Understand the data

In [5]:
store_data.shape

(7501, 20)

In [6]:
store_data.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [8]:
store_data.tail()

0                  1            2               3          4   \
7496    butter         light mayo  fresh bread             NaN        NaN   
7497   burgers  frozen vegetables         eggs    french fries  magazines   
7498   chicken                NaN          NaN             NaN        NaN   
7499  escalope          green tea          NaN             NaN        NaN   
7500      eggs    frozen smoothie  yogurt cake  low fat yogurt        NaN   

             5    6    7    8    9    10   11   12   13   14   15   16   17  \
7496        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7497  green tea  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7498        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7499        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7500        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       18   19  
7496  NaN  NaN  
7497  NaN  NaN  
7498  NaN  NaN  
7499  NaN  NaN  
7500  NaN  NaN

#### Data Preprocessing

The Apriori library requires our dataset to be in the form of a list of lists,
where the whole dataset is a big list and each transaction in the dataset is an inner list within 
the outer big list. Currently we have data in the form of a pandas dataframe. 
To convert our pandas dataframe into a list of lists, execute the following script:

In [12]:
records = []
# Create a list of lists for each transaction record present in store_data.csv

for i in range(0,7501):
    records.append([str(store_data.values[i,j]) for j in range(0,20)\
                   if str(store_data.values[i,j]) !='nan'])
    print([str(store_data.values[i,j]) for j in range(0,20)\
                   if str(store_data.values[i,j]) !='nan'])

['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole weat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie', 'spinach', 'olive oil']
['burgers', 'meatballs', 'eggs']
['chutney']
['turkey', 'avocado']
['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']
['low fat yogurt']
['whole wheat pasta', 'french fries']
['soup', 'light cream', 'shallot']
['frozen vegetables', 'spaghetti', 'green tea']
['french fries']
['eggs', 'pet food']
['cookies']
['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil']
['spaghetti', 'champagne', 'cookies']
['mineral water', 'salmon']
['mineral water']
['shrimp', 'chocolate', 'chicken', 'honey', 'oil', 'cooking oil', 'low fat yogurt']
['turkey', 'eggs']
['turkey', 'fresh tuna', 'tomatoes', 'spaghetti', 'mineral water', 'black tea', 'salmon', 'eggs', 'chicken', 'extra dark chocolate']
['m

['butter', 'burger sauce', 'brownies', 'pancakes']
['butter', 'pancakes', 'whole wheat rice', 'green tea', 'french fries', 'cookies', 'shallot', 'low fat yogurt']
['shrimp', 'chocolate', 'butter', 'cooking oil', 'escalope']
['burgers', 'red wine', 'frozen vegetables', 'pepper', 'spaghetti', 'chocolate', 'milk', 'honey', 'whole wheat rice', 'cooking oil', 'cereals']
['turkey', 'ground beef', 'spaghetti', 'pancakes', 'eggs', 'cake', 'chili', 'pet food', 'chocolate']
['shrimp', 'milk', 'butter', 'chicken', 'salt', 'mayonnaise', 'green tea']
['fresh tuna', 'mineral water', 'fromage blanc', 'honey', 'gums', 'chocolate', 'french fries', 'frozen smoothie', 'sparkling water', 'strawberries', 'green tea']
['chocolate', 'cookies']
['sparkling water', 'light mayo']
['frozen vegetables', 'tomatoes', 'spaghetti', 'chicken', 'french fries', 'escalope']
['turkey', 'burgers', 'butter', 'vegetables mix', 'green grapes', 'pancakes', 'eggs', 'cake', 'barbecue sauce', 'green tea']
['red wine', 'shrimp', '

['light mayo']
['frozen vegetables', 'whole wheat pasta', 'carrots', 'antioxydant juice', 'french fries']
['tomatoes', 'mineral water', 'meatballs', 'chicken', 'pancakes', 'eggs', 'whole wheat rice']
['shrimp', 'frozen vegetables', 'parmesan cheese', 'avocado', 'chicken', 'eggs', 'chocolate', 'frozen smoothie']
['red wine', 'tomatoes', 'spaghetti', 'chocolate', 'chicken', 'bacon', 'pancakes', 'cake', 'rice', 'cooking oil', 'eggplant', 'green tea', 'tea', 'toothpaste']
['spaghetti', 'mineral water', 'soup', 'cider', 'chutney', 'eggs', 'whole wheat rice', 'pet food', 'green tea']
['fresh tuna', 'herb & pepper', 'spaghetti', 'low fat yogurt']
['fresh tuna', 'eggs']
['chocolate', 'escalope', 'pasta', 'honey']
['fresh bread']
['herb & pepper', 'ground beef', 'cooking oil', 'chocolate', 'french fries']
['turkey', 'red wine']
['ham', 'shrimp', 'spaghetti', 'cookies']
['mineral water', 'soup', 'bacon', 'cake', 'light cream', 'eggplant', 'brownies']
['burgers', 'spaghetti', 'eggs', 'french frie

['whole wheat rice', 'frozen smoothie', 'low fat yogurt', 'green tea']
['burgers', 'red wine', 'mineral water', 'cake', 'green tea']
['red wine', 'mineral water', 'soup', 'eggs', 'pet food', 'tomato juice', 'fresh bread']
['grated cheese', 'ground beef', 'pepper', 'spaghetti', 'milk', 'pancakes']
['burgers', 'shrimp', 'frozen vegetables', 'tomatoes', 'light cream', 'extra dark chocolate']
['french fries']
['whole weat flour', 'french fries', 'brownies']
['cake', 'mint green tea']
['spaghetti', 'mineral water', 'cake', 'light cream', 'cooking oil', 'eggplant', 'cottage cheese', 'pancakes']
['ground beef', 'mineral water', 'milk', 'oatmeal', 'chocolate', 'french fries', 'escalope', 'yogurt cake', 'tomato juice']
['olive oil', 'cake', 'eggplant', 'french fries', 'brownies', 'salt', 'fresh bread']
['mineral water', 'french wine', 'green tea']
['mineral water', 'olive oil', 'strong cheese', 'pancakes', 'clothes accessories', 'french fries', 'strawberries']
['spaghetti', 'olive oil']
['grate

['grated cheese', 'whole wheat rice', 'bug spray', 'pasta', 'green tea']
['mineral water', 'milk', 'eggs', 'brownies', 'energy drink']
['energy bar']
['frozen vegetables', 'pancakes', 'low fat yogurt']
['yogurt cake', 'honey']
['cake']
['cereals']
['escalope', 'pasta', 'mushroom cream sauce']
['olive oil', 'light mayo']
['cottage cheese']
['herb & pepper', 'ground beef', 'champagne', 'green tea']
['hot dogs']
['spaghetti', 'french fries', 'energy drink']
['ham', 'shrimp', 'ground beef', 'mineral water', 'avocado', 'milk', 'brownies', 'protein bar']
['mineral water', 'almonds', 'cottage cheese', 'light mayo']
['protein bar']
['red wine', 'candy bars']
['mineral water', 'avocado', 'milk', 'almonds', 'eggs', 'french fries', 'melons']
['herb & pepper']
['vegetables mix', 'escalope']
['tomatoes', 'milk']
['milk', 'salmon']
['eggs']
['mineral water', 'vegetables mix', 'rice', 'light mayo']
['vegetables mix']
['muffins', 'cookies', 'green tea']
['grated cheese', 'spinach', 'eggs', 'cooking oi

['green tea', 'french fries']
['ham', 'spaghetti', 'mineral water', 'olive oil', 'eggs', 'cooking oil', 'french fries', 'green tea']
['shrimp', 'black tea', 'eggs', 'cake']
['shrimp', 'frozen vegetables', 'ground beef', 'spaghetti', 'mineral water', 'olive oil', 'chutney', 'salmon', 'chili', 'cereals', 'strawberries']
['spaghetti', 'hot dogs', 'white wine', 'mayonnaise']
['herb & pepper', 'ground beef']
['white wine']
['green tea']
['frozen vegetables', 'spaghetti', 'olive oil', 'energy bar', 'french wine', 'chicken', 'light cream', 'whole weat flour', 'french fries', 'cottage cheese']
['red wine', 'spaghetti', 'chocolate bread', 'french fries', 'hot dogs', 'low fat yogurt']
['burgers', 'spaghetti', 'avocado', 'cake', 'brownies', 'salt']
['mineral water', 'soup', 'gums', 'extra dark chocolate', 'french fries']
['spaghetti']
['muffins']
['turkey', 'spaghetti', 'olive oil', 'green beans', 'mint green tea', 'chocolate', 'mushroom cream sauce']
['tomatoes', 'whole wheat pasta', 'spaghetti'

['spaghetti', 'salmon', 'pancakes', 'eggs', 'green beans', 'french fries', 'tomato juice', 'zucchini', 'green tea']
['burgers', 'milk', 'champagne', 'mushroom cream sauce', 'green tea']
['herb & pepper', 'soup', 'chicken', 'french fries', 'candy bars']
['spaghetti']
['eggs', 'hot dogs']
['spaghetti', 'salmon', 'nonfat milk', 'eggs', 'carrots']
['spaghetti', 'mineral water', 'meatballs', 'strong cheese']
['ground beef', 'spaghetti', 'mineral water', 'eggs', 'cake', 'chocolate', 'french fries', 'brownies', 'body spray']
['cereals', 'chocolate', 'french fries', 'champagne', 'hot dogs', 'fresh bread', 'cauliflower']
['mineral water']
['oil', 'frozen smoothie', 'white wine']
['turkey', 'eggs']
['frozen vegetables', 'milk', 'green tea', 'low fat yogurt']
['french fries', 'gluten free bar']
['green tea']
['mineral water']
['burgers', 'tomatoes', 'mineral water', 'milk', 'almonds', 'chicken', 'cake', 'french fries', 'green tea']
['burgers', 'frozen vegetables', 'mineral water', 'escalope']
['b

['french fries']
['cookies']
['escalope']
['grated cheese', 'ground beef', 'spaghetti', 'mineral water', 'pancakes', 'cake']
['tomatoes', 'ground beef', 'spaghetti', 'mineral water', 'eggs', 'whole wheat rice', 'cookies', 'low fat yogurt', 'mint']
['whole wheat rice']
['spaghetti', 'soup']
['herb & pepper', 'spaghetti', 'black tea', 'cake', 'cottage cheese', 'yogurt cake']
['cottage cheese']
['pepper', 'mineral water', 'pancakes', 'eggs', 'cooking oil', 'chocolate', 'low fat yogurt']
['grated cheese', 'pancakes', 'blueberries', 'champagne']
['turkey', 'herb & pepper', 'parmesan cheese', 'spaghetti', 'escalope', 'mushroom cream sauce', 'fresh bread', 'low fat yogurt']
['fresh tuna', 'butter', 'chicken', 'green tea', 'chocolate', 'fresh bread']
['frozen vegetables', 'parmesan cheese']
['eggs', 'honey', 'green tea']
['turkey', 'red wine', 'whole wheat pasta', 'pepper', 'spaghetti', 'mineral water', 'soup', 'avocado', 'eggs', 'body spray', 'low fat yogurt']
['mineral water', 'chocolate', '

['french fries']
['spaghetti', 'mineral water', 'avocado', 'almonds', 'french wine', 'oil', 'green tea', 'french fries', 'frozen smoothie', 'melons']
['soup', 'milk']
['tomatoes', 'french fries']
['mint green tea', 'candy bars']
['turkey']
['herb & pepper', 'milk', 'chicken', 'vegetables mix', 'cereals', 'low fat yogurt']
['soup', 'butter', 'chocolate']
['red wine', 'mineral water', 'soup', 'avocado', 'eggs', 'oil', 'corn', 'chocolate bread', 'french fries', 'body spray', 'fresh bread', 'green tea']
['shrimp', 'ground beef', 'pepper', 'mineral water', 'whole wheat rice']
['eggs', 'french fries', 'strawberries']
['escalope']
['grated cheese', 'frozen vegetables']
['chicken', 'muffins', 'cake', 'yogurt cake']
['frozen vegetables', 'tomatoes', 'mineral water', 'cake']
['energy bar']
['grated cheese', 'ground beef', 'whole wheat rice', 'cake']
['burgers', 'grated cheese', 'mineral water', 'cake']
['milk', 'cookies', 'white wine']
['ham', 'pickles', 'spaghetti', 'eggs', 'whole wheat rice', 

['cake', 'french fries']
['ham', 'spaghetti', 'mineral water', 'french wine', 'frozen smoothie', 'cottage cheese']
['mineral water', 'french wine', 'eggs', 'chocolate']
['soda']
['shallot']
['tomatoes', 'mineral water']
['cookies']
['pet food', 'french fries']
['grated cheese', 'milk']
['meatballs']
['oil']
['ham', 'grated cheese', 'chocolate', 'whole wheat rice', 'yogurt cake', 'cauliflower']
['frozen vegetables', 'tomatoes', 'mineral water', 'chicken', 'salmon', 'muffins', 'chocolate', 'green tea']
['french fries', 'cookies']
['french fries']
['ham', 'mineral water', 'milk']
['turkey', 'frozen vegetables', 'spaghetti', 'avocado', 'butter', 'chicken', 'whole wheat rice', 'cake', 'tomato juice']
['olive oil', 'salmon', 'honey']
['frozen vegetables', 'ground beef', 'spaghetti', 'mineral water', 'muffins', 'rice', 'green beans', 'french fries', 'hot dogs']
['burgers', 'low fat yogurt']
['frozen vegetables', 'olive oil', 'tomato juice']
['grated cheese', 'fromage blanc', 'cookies']
['esca

['frozen vegetables', 'eggs', 'pancakes', 'gluten free bar', 'low fat yogurt']
['cookies']
['chicken', 'champagne', 'low fat yogurt']
['chocolate', 'french fries']
['mineral water', 'milk', 'escalope']
['tomatoes', 'spaghetti', 'meatballs', 'strong cheese', 'salmon', 'honey', 'chicken', 'green tea']
['eggs', 'green tea', 'hot dogs', 'cottage cheese', 'brownies']
['mineral water', 'olive oil', 'pancakes', 'chocolate', 'french fries', 'yogurt cake', 'mint']
['red wine', 'whole wheat pasta', 'spaghetti']
['french fries', 'light mayo']
['olive oil', 'cottage cheese', 'fresh bread', 'mint']
['burgers', 'spaghetti', 'shallot', 'mushroom cream sauce']
['pepper', 'spaghetti']
['burgers', 'herb & pepper', 'shrimp', 'frozen vegetables', 'whole wheat pasta', 'ground beef', 'mineral water', 'honey', 'whole wheat rice', 'frozen smoothie', 'escalope', 'magazines']
['ground beef', 'spaghetti', 'mineral water', 'cake', 'magazines', 'green tea']
['mineral water', 'soup', 'french fries', 'mint', 'green 

['muffins', 'green grapes']
['cake', 'french fries']
['chocolate']
['extra dark chocolate', 'french fries', 'green tea']
['whole wheat pasta', 'spaghetti']
['antioxydant juice']
['champagne']
['burgers', 'herb & pepper', 'tomatoes', 'mineral water', 'chicken', 'honey', 'whole wheat rice', 'melons', 'tomato juice']
['olive oil', 'chutney', 'yogurt cake']
['pancakes']
['grated cheese', 'ground beef', 'energy bar', 'honey', 'chicken']
['eggs']
['turkey', 'mineral water', 'milk', 'salmon', 'eggs', 'protein bar']
['hot dogs', 'low fat yogurt', 'candy bars']
['candy bars']
['candy bars']
['french fries', 'cookies']
['escalope', 'cookies']
['vegetables mix', 'french fries']
['shrimp', 'avocado', 'melons']
['pancakes', 'whole wheat rice', 'frozen smoothie', 'brownies']
['frozen vegetables', 'milk', 'chocolate', 'french fries', 'energy drink']
['melons']
['parmesan cheese', 'mineral water', 'muffins', 'chocolate', 'pancakes']
['green tea']
['gums']
['frozen vegetables', 'meatballs', 'pancakes',

['oil', 'cookies']
['frozen vegetables']
['french fries']
['chocolate', 'soup']
['spaghetti', 'yams', 'avocado', 'salmon', 'pancakes', 'light cream', 'bug spray', 'chicken']
['avocado', 'zucchini']
['burgers', 'spaghetti', 'green grapes', 'cake', 'frozen smoothie', 'cottage cheese']
['flax seed', 'cake', 'pancakes']
['whole wheat rice', 'french fries']
['tomatoes']
['chocolate', 'french fries', 'frozen smoothie', 'hot dogs', 'low fat yogurt']
['green tea']
['burgers', 'sandwich']
['mineral water', 'salmon', 'chocolate']
['french fries']
['mineral water', 'milk', 'salmon']
['cake', 'pancakes', 'energy drink']
['shrimp', 'frozen vegetables', 'mineral water', 'soup', 'milk', 'pancakes', 'chocolate', 'french fries', 'green tea']
['muffins']
['mineral water', 'butter', 'spinach', 'extra dark chocolate', 'antioxydant juice', 'sparkling water', 'light mayo']
['mineral water', 'avocado', 'milk', 'honey']
['turkey', 'parmesan cheese']
['shrimp', 'oil', 'champagne']
['milk', 'chocolate', 'french

['frozen vegetables', 'green grapes']
['turkey', 'green tea']
['herb & pepper', 'ground beef', 'pancakes', 'burger sauce', 'french fries']
['soup', 'muffins', 'rice', 'cereals']
['chocolate', 'cookies', 'light mayo']
['milk', 'low fat yogurt']
['ground beef', 'spaghetti', 'carrots', 'chocolate', 'frozen smoothie']
['tomatoes', 'energy bar', 'vegetables mix', 'cottage cheese', 'pancakes', 'fresh bread']
['chocolate', 'mineral water']
['fresh tuna', 'mineral water', 'french wine', 'cereals']
['yogurt cake']
['protein bar', 'tomato juice', 'fresh bread', 'honey']
['eggs', 'cooking oil', 'eggplant']
['burgers', 'fresh tuna', 'milk', 'chicken', 'champagne', 'asparagus', 'white wine']
['green tea', 'frozen smoothie']
['turkey', 'mineral water']
['burgers', 'energy bar', 'whole wheat rice', 'cake']
['mineral water', 'chocolate', 'milk', 'pancakes', 'eggs']
['avocado', 'energy bar']
['tomatoes', 'spaghetti', 'mineral water', 'meatballs', 'salt']
['burgers', 'spaghetti', 'milk', 'carrots', 'cho

['eggs', 'cookies']
['cookies']
['tomatoes', 'meatballs', 'chicken', 'eggs']
['tomatoes', 'chocolate', 'eggs', 'pancakes']
['eggs', 'whole wheat rice']
['herb & pepper']
['nonfat milk', 'escalope']
['french fries', 'yogurt cake']
['cookies']
['frozen vegetables']
['yams', 'eggplant']
['eggs']
['cookies']
['chocolate', 'pancakes', 'light cream', 'chili', 'extra dark chocolate']
['whole wheat pasta', 'french fries', 'brownies']
['ground beef', 'cookies', 'toothpaste', 'green tea']
['ham', 'whole wheat rice', 'french fries', 'champagne', 'cauliflower']
['whole wheat rice']
['shrimp', 'milk', 'butter', 'cake', 'chocolate bread', 'green tea']
['mineral water', 'sparkling water', 'yogurt cake', 'energy drink', 'light mayo']
['french fries', 'escalope', 'pasta', 'mushroom cream sauce', 'green tea']
['ground beef', 'spaghetti', 'yams', 'eggs', 'cookies', 'mushroom cream sauce']
['frozen vegetables', 'french fries']
['ground beef', 'spaghetti', 'mineral water', 'chocolate', 'salmon', 'chicken',

['brownies', 'pancakes']
['milk', 'champagne']
['muffins']
['mineral water']
['burgers', 'shrimp', 'frozen vegetables', 'spaghetti', 'soup', 'cake', 'french fries', 'pancakes', 'tomato juice']
['cottage cheese', 'protein bar']
['burgers', 'eggs', 'french fries']
['champagne']
['turkey', 'mineral water', 'fromage blanc', 'muffins', 'eggplant', 'tomato juice', 'magazines']
['cake', 'cooking oil', 'light mayo']
['strawberries']
['burgers', 'french fries']
['chocolate', 'grated cheese', 'frozen smoothie', 'cookies', 'brownies']
['energy drink']
['sparkling water']
['mineral water', 'flax seed']
['turkey', 'herb & pepper', 'red wine', 'parmesan cheese', 'mineral water', 'olive oil', 'bacon', 'salmon', 'green tea', 'french fries', 'champagne', 'cottage cheese', 'tomato juice']
['grated cheese', 'shrimp', 'ground beef', 'mineral water', 'avocado', 'milk', 'nonfat milk', 'cooking oil', 'carrots', 'pancakes', 'light mayo']
['frozen vegetables', 'mineral water', 'pancakes', 'low fat yogurt']
['s

In [30]:
#store_data.values

#### Applying Apriori

In [31]:
records

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spagh

In [32]:
type(records)

list

### Steps involved in Apriori Algorithm
1. Set a minimum value for support and confidence. 
This means that we are only 
interested in finding rules for the items that have certain default existence 
(e.g. support) and have a minimum value for co-occurrence 
with other items (e.g. confidence).

2. Extract all the subsets having higher value of support than minimum threshold.

3. Select all the rules from the subsets with confidence value higher than minimum threshold.

#### Parameters
##### min_support - Refers to minimum popularity of an item.
##### min_confidence - Tells us how much more likely it is to buy item B given that item A are bought.
##### min_lift - Tells us that the likelihood of buying item A and item B together is x times more than the likelihood of just buying the item A. 


In [38]:
# Running the Apriori algorithm on transactions data with minimum support, confidence and list values and covert the output to a list
association_results = apriori(records, min_support=0.0046, min_confidence=0.2, min_lift = 2, min_length=3)
association_results = list(association_results)

In [39]:
print(len(association_results))

135


In [40]:
print(association_results)

[RelationRecord(items=frozenset({'burgers', 'almonds'}), support=0.005199306759098787, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.25490196078431376, lift=2.923577382023146)]), RelationRecord(items=frozenset({'ham', 'burgers'}), support=0.005599253432875617, ordered_statistics=[OrderedStatistic(items_base=frozenset({'ham'}), items_add=frozenset({'burgers'}), confidence=0.21105527638190955, lift=2.420681388594348)]), RelationRecord(items=frozenset({'milk', 'cereals'}), support=0.007065724570057326, ordered_statistics=[OrderedStatistic(items_base=frozenset({'cereals'}), items_add=frozenset({'milk'}), confidence=0.2746113989637306, lift=2.119197637476279)]), RelationRecord(items=frozenset({'tomato sauce', 'chocolate'}), support=0.005065991201173177, ordered_statistics=[OrderedStatistic(items_base=frozenset({'tomato sauce'}), items_add=frozenset({'chocolate'}), confidence=0.3584905660377358, lift=2.1879883936932925)

In [41]:
print(association_results[1])

RelationRecord(items=frozenset({'ham', 'burgers'}), support=0.005599253432875617, ordered_statistics=[OrderedStatistic(items_base=frozenset({'ham'}), items_add=frozenset({'burgers'}), confidence=0.21105527638190955, lift=2.420681388594348)])


* Each RelationRecord  reflects all rules associated with a specific itemset (items) that has relevant rules. Support (support ), given that it’s simply a count of appearances of those items together, is the same for any rules involving those items, and so only appears once per RelationRecord. The ordered_statistic  reflects a list of all rules that met our min_confidence  and min_lift  requirements (parameterized when we called apriori() ). Each OrderedStatistic  contains the antecedent (items_base) and consequent (items_add) for the rule, as well as the associated confidence  and lift .

In [42]:
association_results[0]

RelationRecord(items=frozenset({'burgers', 'almonds'}), support=0.005199306759098787, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.25490196078431376, lift=2.923577382023146)])

In [43]:
association_results[0][0]

frozenset({'almonds', 'burgers'})

In [44]:
association_results[0][1]

0.005199306759098787

In [45]:
association_results[0][2]

[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.25490196078431376, lift=2.923577382023146)]

In [46]:
item=association_results[0]

In [47]:
print(item.items)
print(item.support)
print(item.ordered_statistics)

frozenset({'burgers', 'almonds'})
0.005199306759098787
[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.25490196078431376, lift=2.923577382023146)]


In [55]:
# Print Support, Lift and Confidence for the resultant transactions
print("support = ",item[0])
print("Confidence =", item[2][0][2])
print("lift =", item[2][0][3])

support =  frozenset({'burgers', 'almonds'})
Confidence = 0.25490196078431376
lift = 2.923577382023146


In [92]:
for item in association_results:

    # first index of the inner list
    # Contains base item and add item
    pair = item[0]
    items = [x for x in pair]
    print(items)
    print("Rule: " + items[0] + " -> " + items[1])

    # second index of the inner list
    print("Support: " + str(item[1]))

    # third index of the list located at 0th
    # of the third index of the inner list

    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("=====================================")

['almonds', 'burgers']
Rule: almonds -> burgers
Support: 0.005199306759098787
Confidence: 0.25490196078431376
Lift: 2.923577382023146
['burgers', 'ham']
Rule: burgers -> ham
Support: 0.005599253432875617
Confidence: 0.21105527638190955
Lift: 2.420681388594348
['cereals', 'milk']
Rule: cereals -> milk
Support: 0.007065724570057326
Confidence: 0.2746113989637306
Lift: 2.119197637476279
['chocolate', 'tomato sauce']
Rule: chocolate -> tomato sauce
Support: 0.005065991201173177
Confidence: 0.3584905660377358
Lift: 2.1879883936932925
['mushroom cream sauce', 'escalope']
Rule: mushroom cream sauce -> escalope
Support: 0.005732568990801226
Confidence: 0.3006993006993007
Lift: 3.790832696715049
['escalope', 'pasta']
Rule: escalope -> pasta
Support: 0.005865884548726837
Confidence: 0.3728813559322034
Lift: 4.700811850163794
['mineral water', 'extra dark chocolate']
Rule: mineral water -> extra dark chocolate
Support: 0.005732568990801226
Confidence: 0.47777777777777775
Lift: 2.0043686303753416


**Rule Interpretation:**

Rule: ham -> burgers
* Support: 0.005599253432875617 - Popularity of both burgers and ham bought together is 0.005599253432875617
* Confidence: 0.21105527638190955 - The probability of ham being bought with burger is 0.21 (21%). (Out of all the transactions containing burger, 21% of the transactions are likely to contain ham as well)
* Lift:  2.420681388594348 - It is 2.4 times likely of Burger and ham being bought together than just ham alone. 